File: rate_limit/test/rate_limit-tests.py
"""
Copyright June 25, 2020 Richard Koshak

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
"""
from community.rate_limit import RateLimit
import time
from core.log import logging, LOG_PREFIX

log = logging.getLogger("{}.TEST.Latch".format(LOG_PREFIX))
func_called = False

def test():
    global func_called
    func_called = True

try:
    log.info("Starting rate_limit tests")
    rate_limit = RateLimit()

    rate_limit.run(test, secs=2)
    assert func_called, "Test1: function was not called"

    func_called = False
    rate_limit.run(test, secs=2)
    assert not func_called, "Test2: function was called"

    time.sleep(2)
    rate_limit.run(test, secs=2)
    assert func_called, "Test3: function was not called"

except AssertionError:
    import traceback
    log.error("Exception: {}".format(traceback.format_exc()))
else:
    log.info("Latch tests passed!")

File: tests/rateLimitTests.js
var {rateLimit, testUtils} = require('openhab_rules_tools');

var logger = log('rules_tools.RateLimit Tests');

var funcCalled = false;

var test = () => funcCalled = true;

var rl = new rateLimit.RateLimit();

logger.info('Starting RateLimit tests');

// Test 1: called first time
rl.run(test, 2);
testUtils.assert(funcCalled, 'Test 1: First run function was not called');

// Test 2: calling after 2 second wait func is called
funcCalled = false;
testUtils.sleep(2);
rl.run(test, '2s');
testUtils.assert(funcCalled, 'Test 2: First run function was not called');
funcCalled = false
testUtils.sleep(2000);
rl.run(test, 2);
testUtils.assert(funcCalled, 'Test 2: Second run was not calle after two seconds');

// Test 3: calling before wait func is not called
funcCalled = false;
testUtils.sleep(2);
rl.run(test, '2s');
testUtils.assert(funcCalled, 'Test 3: First run function was not called');
funcCalled = false;
rl.run(test, '2s');
testUtils.assert(!funcCalled, 'Test 3: Second run function was called before wait time');

logger.info('All tests have passed');